### PR TITLE
MIXEDARCH-169: Introducing Azure multi-arch compute nodes and other fixes to ipi-install-heterogeneous

### PR DESCRIPTION
**Pull request by aleskandro**
This PR adds support for the Azure jobs using the multi payload to add workers with a different architecture than the control plane and fixes the ipi-install-heterogeneous to wait for the additional workers to become ready before proceeding with the next jobs. 

Finally, it simplifies the step by (a) converging all the yaml/json processing logic to a uniq version of `yq` and (b) remove some unneeded logic.

Summary of the changes:

- Introducing the logic to add worker nodes with a different architecture than the control plane on Azure clusters using the multi payload
- Removes un-needed dependencies on other versions of yq/jq by converging any yaml/json processing to yq-v4
- Fixes the creation of the MachineSet: the additional nodes must be ready before starting the next steps
- Fixes PATH assignment when yq-v4 already exists
- Removes the data collection script as the previous installation steps already handle that (/cc @deepsm007 can you double check and confirm?)
- Use the upi-installer image to have Cloud provider tools in place
- Adds a job in the QE configs to execute the openshift-tests-private suite against clusters with multi-arch compute nodes.

Refers [MIXEDARCH-169](https://issues.redhat.com//browse/MIXEDARCH-169)

#### Regression test for AWS (4.12) e2e test suite

/pj-rehearse periodic-ci-openshift-multiarch-master-nightly-4.12-ocp-e2e-serial-aws-ovn-heterogeneous

Installation and scale-out succeed initially at:

https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_release/35101/rehearse-35101-periodic-ci-openshift-multiarch-master-nightly-4.12-ocp-e2e-serial-aws-ovn-heterogeneous/1610239241483194368

#### New job config execution (Azure IPI + ARM64 workers, openshift-tests-private)

/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.13-multi-nightly-azure-ipi-baselinecaps-v413-p2-f14

Installation and scale-out succeed initially at:

https://qe-private-deck-ci.apps.ci.l2s4.p1.openshiftapps.com/view/gs/qe-private-deck/pr-logs/pull/openshift_release/35101/rehearse-35101-periodic-ci-openshift-openshift-tests-private-release-4.13-multi-nightly-azure-ipi-baselinecaps-v413-p2-f14/1610259061004046336

I force pushed another commit after those runs to remove some lines to ease the debug

/cc @Prashanth684 @SaravanaStorageNetwork  @deepsm007

### CODE REVIEW

**Comment by aleskandro** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.13-multi-nightly-azure-ipi-baselinecaps-v413-p2-f14

**Comment by aleskandro** [comment]
/pj-rehearse periodic-ci-openshift-multiarch-master-nightly-4.12-ocp-e2e-serial-aws-ovn-heterogeneous

**Comment by aleskandro** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.13-multi-nightly-azure-ipi-baselinecaps-v413-p2-f14

File: ci-operator/step-registry/ipi/install/heterogeneous/ipi-install-heterogeneous-commands.sh [hunk 1]
```diff
@@ -6,98 +6,133 @@ set -o pipefail
...
+    fi
+    sleep "${period}m"
+    ready_nodes=$(get_ready_nodes_count)
+    if [ x"${ready_nodes}" == x"${expected_nodes}" ]; then
+        echo "[INFO] Found ${ready_nodes}/${expected_nodes} ready nodes, continuing..."
+        return 0
+    fi
+    echo "[INFO] - ${expected_nodes} ready nodes expected, found ${ready_nodes}..." \
+      "Waiting ${period}min before retrying (timeout in $(( (max_retries - i) * (period) ))min)..."
+    (( max_retries -- ))
```
Comment (Prashanth684) [inline ci-operator/step-registry/ipi/install/heterogeneous/ipi-install-heterogeneous-commands.sh:34]: if it does timeout , do we collect must-gather ?

Comment (aleskandro) [reply ci-operator/step-registry/ipi/install/heterogeneous/ipi-install-heterogeneous-commands.sh:34]: good point, let me check. We might need to delete some file in the SHARED_DIR.

Comment (aleskandro) [reply ci-operator/step-registry/ipi/install/heterogeneous/ipi-install-heterogeneous-commands.sh:34]: regardless of that, I don't need the line you are commenting on :D

File: ci-operator/step-registry/ipi/install/heterogeneous/ipi-install-heterogeneous-commands.sh [hunk 2]
```diff
@@ -6,98 +6,133 @@ set -o pipefail
...
+    echo "Waiting for the VHD blob copy to conclude... (timeout in $(( 15 - i )) minutes)"
+  done
+  echo "The VHD image is now available. Creating the image version..."
+  az sig image-version create --resource-group "${rg_name}" \
+    --gallery-name "${gallery_name}" --gallery-image-definition "${image_name}" \
+    --gallery-image-version 1.0.0 --target-regions "${region}" \
+    --os-vhd-uri "${storage_blob_url}" --os-vhd-storage-account "${sa_name}"
+  echo "The image version for the ${ADDITIONAL_WORKER_ARCHITECTURE} workers has been created... "
+  echo "Patching the MachineSet..."
+  resource_id="/resourceGroups/${rg_name}/providers/Microsoft.Compute/galleries/${gallery_name}/images/${image_name}/versions/1.0.0"
```
Comment (Prashanth684) [inline ci-operator/step-registry/ipi/install/heterogeneous/ipi-install-heterogeneous-commands.sh:124]: you can have the image version be any string and just reference it as "latest"..that way you won't have to set a specific version. see https://github.com/openshift/installer/blob/master/pkg/asset/machines/azure/machines.go#L120

File: ci-operator/step-registry/ipi/install/heterogeneous/ipi-install-heterogeneous-commands.sh (hunk 1 above)
Comment (aleskandro) [reply ci-operator/step-registry/ipi/install/heterogeneous/ipi-install-heterogeneous-commands.sh]: removed the line and add this other to allow the `gather` chain to do its job.

Comment (aleskandro) [reply ci-operator/step-registry/ipi/install/heterogeneous/ipi-install-heterogeneous-commands.sh]: https://github.com/openshift/release/pull/35101/files#diff-9e75860c660730e2cb77472d634a02aa6da64e7bac073d6cbfb7daf4133188a1R138

File: ci-operator/step-registry/ipi/install/heterogeneous/ipi-install-heterogeneous-commands.sh (hunk 2 above)
Comment (aleskandro) [reply ci-operator/step-registry/ipi/install/heterogeneous/ipi-install-heterogeneous-commands.sh]: done and the version is now set similarly to the one the installer sets: (41x.yz.YYYYMMDD)

https://github.com/openshift/release/pull/35101/files#diff-9e75860c660730e2cb77472d634a02aa6da64e7bac073d6cbfb7daf4133188a1R119

**Comment by aleskandro** [comment]
The failures in the jobs are unrelated. Triggering another few runs with the review changes.

The others were good at installation time and, in particular, in the step affected by this PR.

/pj-rehearse periodic-ci-openshift-multiarch-master-nightly-4.12-ocp-e2e-aws-ovn-heterogeneous-upgrade

/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.13-multi-nightly-azure-ipi-baselinecaps-v413-p2-f14

**Comment by aleskandro** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.13-multi-nightly-azure-ipi-baselinecaps-v413-p2-f14

File: ci-operator/step-registry/ipi/install/heterogeneous/ipi-install-heterogeneous-commands.sh (hunk 1 above)
Comment (Prashanth684) [reply ci-operator/step-registry/ipi/install/heterogeneous/ipi-install-heterogeneous-commands.sh]: looks good!

Comment (aleskandro) [reply ci-operator/step-registry/ipi/install/heterogeneous/ipi-install-heterogeneous-commands.sh]: As discussed offline, we don't need it. must-gather is taken in any case. The exit code in the install-status.txt file will feed the junits to report the installation status. As far as we keep the scale-out of the additional workers as a day2 operation, we're good to go without it. Resolving the thread

**Comment by aleskandro** [comment]
/unhold

/pj-rehearse ack

The commit made a few minutes ago doesn't need retesting the rehearsals (mostly removed a non-needed statement). 

/cc @MayXuQQ 

/assign @liangxia

File: ci-operator/step-registry/ipi/install/heterogeneous/ipi-install-heterogeneous-commands.sh [hunk 3]
```diff
@@ -6,98 +6,134 @@ set -o pipefail
...
+        echo "[INFO] Found ${ready_nodes}/${expected_nodes} ready nodes, continuing..."
+        return 0
+    fi
+    echo "[INFO] - ${expected_nodes} ready nodes expected, found ${ready_nodes}..." \
+      "Waiting ${period}min before retrying (timeout in $(( (max_retries - i) * (period) ))min)..."
+  done
+}
+
+# Make sure yq-v4 is installed
 if [ ! -f /tmp/yq-v4 ]; then
```
Comment (MayXuQQ) [inline ci-operator/step-registry/ipi/install/heterogeneous/ipi-install-heterogeneous-commands.sh:38]: upi-installer image installed yq 
https://github.com/openshift/installer/blob/master/images/installer/Dockerfile.upi.ci
https://github.com/openshift/installer/blob/master/images/installer/Dockerfile.upi.ci.rhel8

Comment (aleskandro) [reply ci-operator/step-registry/ipi/install/heterogeneous/ipi-install-heterogeneous-commands.sh:38]: it's another version. The upgrade of yq is tracked elsewhere

File: ci-operator/step-registry/ipi/install/heterogeneous/ipi-install-heterogeneous-commands.sh [hunk 4]
```diff
@@ -6,98 +6,134 @@ set -o pipefail
...
+  vhd_name=$(basename "${vhd_url}")
+  infra_id=$(yq-v4 '.infraID' < "${SHARED_DIR}"/metadata.json)
+  rg_name="${infra_id}-rg"
+  sa_name=$(az storage account list -g "${rg_name}" | yq-v4 '.[] | select(.name == "cluster*").name')
+  AZURE_STORAGE_KEY=$(az storage account keys list -g "${rg_name}" --account-name "${sa_name}" --query "[0].value" -o tsv)
+  export AZURE_STORAGE_KEY
+  az storage blob copy start --account-name "${sa_name}" \
+      --destination-blob "${vhd_name}" --destination-container vhd --source-uri "$vhd_url"
+  gallery_name=$(az sig list -g "${rg_name}" | yq-v4 '.[].name')
+  image_name="${infra_id}-gen2-${ADDITIONAL_WORKER_ARCHITECTURE}"
```
Comment (MayXuQQ) [inline ci-operator/step-registry/ipi/install/heterogeneous/ipi-install-heterogeneous-commands.sh:101]: @aleskandro  with the image_name test pass ?

Comment (aleskandro) [reply ci-operator/step-registry/ipi/install/heterogeneous/ipi-install-heterogeneous-commands.sh:101]: yes, it does. You can find the history of the installation jobs with the successful scale-out on Azure starting from the links in the description.

File: ci-operator/step-registry/ipi/install/heterogeneous/ipi-install-heterogeneous-commands.sh [hunk 5]
```diff
@@ -6,98 +6,134 @@ set -o pipefail
...
+  region=$(az group show --name "${rg_name}" | yq-v4 '.location')
+  for i in $(seq 1 15) max; do
+    [ "$i" == max ] && { echo "Timeout exceeded while waiting for the VHD blob copy to conclude. Failing..."; exit 3; }
+    sleep 60
+    [ X"$(az storage blob show --container-name vhd --name "${vhd_name}" --account-name "${sa_name}" \
+      -o tsv --query properties.copy.status)" == X"success" ] && break
+    echo "Waiting for the VHD blob copy to conclude... (timeout in $(( 15 - i )) minutes)"
+  done
+  echo "The VHD image is now available. Creating the image version..."
+  az sig image-version create --resource-group "${rg_name}" \
```
Comment (MayXuQQ) [inline ci-operator/step-registry/ipi/install/heterogeneous/ipi-install-heterogeneous-commands.sh:117]: still need the image version ? 
https://github.com/openshift/installer/pull/6684/files 
https://github.com/openshift/installer/pull/6684#issuecomment-1361687003 
and https://github.com/openshift/release/pull/34978/files usage of image version parameter

Comment (aleskandro) [reply ci-operator/step-registry/ipi/install/heterogeneous/ipi-install-heterogeneous-commands.sh:117]: https://qe-private-deck-ci.apps.ci.l2s4.p1.openshiftapps.com/view/gs/qe-private-deck/pr-logs/pull/openshift_release/35101/rehearse-35101-periodic-ci-openshift-openshift-tests-private-release-4.13-multi-nightly-azure-ipi-baselinecaps-v413-p2-f14/1610338624946049024

we use `latest` later in the reference.

**Pull request merged**